api/prometheus/v1/api_test.go

```diff
@@ -194,6 +198,30 @@ func TestAPIs(t *testing.T) {
 			},
 			err: fmt.Errorf("some error"),
 		},
+		{
+			do:     doQuery("2", testTime),
+			inRes:  "some body",
+			inCode: 500,
+			inErr: &Error{
+				Type:   ErrBadResponse,
```

Comment:
Right now `ErrBadResponse` is [returned for all responses with codes other than 2xx]( https://github.com/prometheus/client_golang/blob/93edea3720f32de482abe8b4a6c1d9077d1e9981/api/prometheus/v1/api.go#L471-L473). And technically `4xx` are not server errors but client errors, so claiming a server error on a `404` for example would be counter-intuitive. Do you want me to treat `5xx` errors specially? Otherwise we'd need a generic const name for "a response that is not 2xx"...

Reply:
yeah you can change it to return the error type based on the error codes.

Reply:
Done, PTAL.

I also improved the table tests a bit to make them easier to debug and understand.